Closes #4 - Added automatic feature recording to annotate_img

msekatchev opened the pull request:
Closes #4 
annotate_img: You now select a PMT ID and starting feature number, and can record incrementing feature numbers without reentering the PMT ID.

msekatchev committed:
Update annotate.py

Line 36: added cv2.WINDOW_NORMAL flag to allow the user to resize the window.
Working on zoom capability for Windows.
cv2.namedWindow('image',cv2.WINDOW_NORMAL)

msekatchev committed:
Update annotate.py

msekatchev committed:
Added 2D List for recording pixel coordinates

Added a 2D list for recording pixel coordinates. 
Currently only tested with annotate_img.
The plan is to have this list saved to a file with each line having the x and y pixel coordinates preceded by a PMT feature #.

msekatchev committed:
annotate_img now saves a text file with PMT feature coordinates

- Worked on annotate_img function (will work on extending modifications to the annotate_dir function).
- The function now creates a text file. Each line contains a PMT feature ID and its pixel coordinates on the image. The text file has the same name as the image.

see README.txt for instructions.

msekatchev committed:
Update README.txt

msekatchev committed:
annotate_dir now saves a text file with PMT feature coordinates

msekatchev committed:
Update README.txt

msekatchev committed:
Cleaned up the code

This will be the last version where annotate_dir saves to a single file for all the images combined.

msekatchev committed:
Detailed instructions for installing prerequisites and running script

Initial draft version.
Will be modified to add annotate_dir instructions and change subfolder system.

msekatchev committed:
Added automatic feature recording to annotate_img

annotate_img: You now select a PMT ID and starting feature number, and can record incrementing feature numbers without reentering the PMT ID.

pdeperio commented:
thanks, good start! now few more iterations to make this feature production-ready (e.g. optimizing user experience and dummy-proofing):

1. If I make a mistake, can I delete the point/marker, say by pressing `d`? I know I can just click to make a new point then save that one with `r`, but just so we don't crowd the mask or get confused later with unsaved points. 

2. Can I increment the feature ID, say with `b` (back) and `n` (next), instead of just `x` for next/skip? This would let me edit previous points if I needed to. (This might require saving data in an array in memory first, before writing to file, so the array can be modified then written at the very end.)

msekatchev committed:
Added zero padding around feature number

msekatchev committed:
Closes #1 - Streamlined directory paths for annotate_dir()

Closes #1:
annotate_dir now uses the following directory scheme, where "images" is the inputted img_dir:
images
-> .JPGs
images_masks
-> saved masks
images_texts
-> saved text file (will work on making a single file per image)

Directories images_texts and images_masks will be created automatically if they don't exist.
Need to add overwrite protection for masks and text file(s).

pdeperio commented on annotate.py:
```diff
@@ -106,7 +106,10 @@ def annotate_img(img_path, size1, size2, initials) :
                 print("Recording ", pmtID, "-",startingVal, " ",ix, "x  ", iy,"y\n")
-                file.write("%s\t%s-%d\t%d\t%d\t%s\n" %(filename,pmtID,startingVal, ix, iy,initials))
+                if(startingVal<10 or startingVal == 0):
+                    file.write("%s\t%s-0%d\t%d\t%d\t%s\n" %(filename,pmtID,startingVal, ix, iy,initials))
```
[Here's a more programmatic way](https://stackoverflow.com/questions/339007/how-to-pad-zeroes-to-a-string) (non-hardcoding) to zero-pad a number.

Could you also pad the PMT ID? i.e. `%05d`

In general, try to find ways to avoid hardcoding numbers, reducing the number of `if` blocks (which the above method fixes), and instances of large copy/paste.

For example `"%s\t%s-%02d\t%d\t%d\t%s\n" %(filename,pmtID,startingVal, ix, iy,initials)` can be defined as one variable before, then used everywhere below.

pdeperio replied:
fixes #5

pdeperio replied:
just testing if this ^ works later when merging

pdeperio reviewed (changes requested):
Few improvements before merging...

pdeperio commented on annotate.py:
```diff
@@ -180,7 +183,7 @@ def annotate_img(img_path, size1, size2, initials) :
 #       filename - The name of the .txt file that will be created.
 #Set up for directory of images with file structure for image segmentation
-def annotate_dir(img_dir, dataset, subset, size1, size2,initials,filename) :
+def annotate_dir(img_dir, size1, size2,initials,filename) :
```
Since we're moving to a per-image output structure, `filename` can be removed here, and instead a text file for each image saved below.

For a future PR,  this function should in principle just be a loop calling `annotate_img` (or some new function that consolidates the copy/pasted code between the two functions).

msekatchev replied:
Yeap, that's issue #2, I will change the code to have `annotate_dir` call `annoate_img` first.

pdeperio replied:
right, forgot. make a new PR for that then

pdeperio commented on annotate.py:
```diff
@@ -194,9 +197,14 @@ def annotate_dir(img_dir, dataset, subset, size1, size2,initials,filename) :
+        os.mkdir(text_save_path)
+    if not os.path.exists(mask_save_path):
+        os.mkdir(mask_save_path)
+    save_path = os.path.join(img_dir+"_texts",filename+".txt")
```
See first comment above. This should go into the loop below, creating a new `.txt` file per image.

pdeperio reviewed (changes requested):
Few more requests before merging...

msekatchev committed:
annotate_dir now calls annotate_img and saves to separate text files

Closes #2.

msekatchev committed:
Added zero padding around PMT-ID and Feature #

Closes #5.

msekatchev committed:
Added zero padding around PMT-ID and Feature #

closes #5.

msekatchev committed:
Merge pull request #2 from msekatchev/msekatchev-issue-5

Added zero padding around PMT-ID and Feature #

msekatchev committed:
Update README.txt
